addons/website/static/src/snippets/s_image_gallery/options.js

```diff
@@ -455,6 +496,25 @@ options.registry.gallery = options.Class.extend({
         $container.empty().append($content);
         return $container;
     },
+    /**
+     * Call mode while ensuring that all images are loaded.
+     *
+     * @see this.selectClass for parameters
+     * @returns {Promise}
+     */
+    _modeWithImageWait(previewMode, widgetValue, params) {
+        // TODO Remove in master.
+        let promise;
+        this._masonryAwaitImages = true;
+        try {
+            this.mode(previewMode, widgetValue, params);
+            promise = this._masonryAwaitImagesPromise;
+        } finally {
+            this._masonryAwaitImages = false;
+            this._masonryAwaitImagesPromise = undefined;
+        }
+        return promise || Promise.resolve();
```

Comment:
I'm not sure I understand why this line is needed, could you explain?

Reply:
TL;DR I'd rather not have methods rely on their global call context to maintain their own consistency.

By not using `finally`, if `mode` failed, the patch would remain applied for calls to `mode` that do not run through `_modeWithImageWait`.
Indeed in the current code base `mode` cannot fail and if it fails, there is no recovery mechanism within the website builder - which would ultimately lead to the destruction of the instance and therefore no further call to its `mode` method.
However, given that we introduce this hack because there could be unforeseenable customizations, I consider that not removing the patch could be introducing a side-effect.
If a custom layout is introduced and has the same kind of issue as masonry, this method might end up being used in future customizations. Who knows what might come up... a server-side defined layout, a "shuffle images" option... possibly introducing the need for catching errors on the async layout calculation call. In such a case, it makes no sense for the caller code to have to handle such a cleanup.
Even if hypothetical and very unlikely, I think the method should ensure to never break the behavior of other methods.
I would have preferred to use a Python `with resource:` or a Java `try(resource) {...}` syntax, which would not raise the question of the need to cleanup.